File: mc-node-automesh/src/registry-client.ts
import { PublishPayload } from "./types.js";

/**
 * RegistryClient handles publishing tools to the MCPRPC registry.
 */
export class RegistryClient {
  private registryUrl: string;

  constructor(registryUrl: string) {
    this.registryUrl = registryUrl;
  }

  /**
   * Publishes a tool to the registry.
   */
  async publish(payload: PublishPayload): Promise<void> {
    const url = new URL("/register", this.registryUrl).href;

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        const text = await response.text().catch(() => "");
        throw new Error(`Registry responded with status ${response.status}: ${text}`);
      }
      console.warn(
        `event=registry_publish ok=true name=${payload.name} service_name=${payload.service_name} mesh_id=${payload.mesh_id}`
      );
    } catch (error: any) {
      console.warn(
        `event=registry_publish ok=false name=${payload.name} service_name=${payload.service_name} mesh_id=${payload.mesh_id} error=${error.message}`
      );
      throw new Error(`Failed to publish to registry: ${error.message}`);
    }
  }

  async heartbeat(payload: {
    service_name: string;
    mesh_id: string;
    runtime?: string;
    health: string;
    tools?: string[];
    heartbeat_interval_s?: number;
    registrations?: PublishPayload[];
  }): Promise<void> {
    const url = new URL("/heartbeat", this.registryUrl).href;

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        const text = await response.text().catch(() => "");
        throw new Error(`Registry responded with status ${response.status}: ${text}`);
      }
      console.warn(
        `event=registry_heartbeat ok=true service_name=${payload.service_name} mesh_id=${payload.mesh_id} health=${payload.health}`
      );
    } catch (error: any) {
      console.warn(
        `event=registry_heartbeat ok=false service_name=${payload.service_name} mesh_id=${payload.mesh_id} health=${payload.health} error=${error.message}`
      );
      throw new Error(`Failed to send heartbeat to registry: ${error.message}`);
    }
  }
}

File: mc-java-automesh/src/main/java/io/mcprpc/automesh/RegistryClient.java
package io.mcprpc.automesh;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

final class RegistryClient {
    private static final ObjectMapper MAPPER = SchemaUtils.MAPPER;
    private final HttpClient httpClient;
    private final String registryUrl;

    RegistryClient(String registryUrl) {
        this.registryUrl = registryUrl == null ? "" : registryUrl;
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(5))
            .version(HttpClient.Version.HTTP_1_1)
            .build();
    }

    void publish(Map<String, Object> payload) throws IOException, InterruptedException {
        if (registryUrl.isBlank()) {
            return;
        }
        send("/register", payload);
    }

    void heartbeat(Map<String, Object> payload) throws IOException, InterruptedException {
        if (registryUrl.isBlank()) {
            return;
        }
        send("/heartbeat", payload);
    }

    Map<String, Object> buildPublishPayload(String serviceName, String meshId, String runtime, String mcpTransport, String endpoint, ToolMetadata metadata) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("name", metadata.name());
        payload.put("mesh_id", meshId);
        payload.put("service_name", serviceName);
        payload.put("runtime", runtime);
        payload.put("transport", "mcp");
        payload.put("mcp_transport", mcpTransport);
        payload.put("endpoint", endpoint);
        payload.put("description", metadata.description());
        payload.put("inputSchema", metadata.inputSchema());
        payload.put("outputSchema", metadata.outputSchema());
        payload.put("acl", metadata.acl());
        payload.put("cost", metadata.cost());
        payload.put("tags", metadata.tags());
        payload.put("version", metadata.version());
        payload.put("health", "healthy");
        return payload;
    }

    Map<String, Object> buildHeartbeatPayload(String serviceName, String meshId, String runtime, String health, int heartbeatIntervalSeconds, List<String> tools, List<Map<String, Object>> registrations) {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("service_name", serviceName);
        payload.put("mesh_id", meshId);
        payload.put("runtime", runtime);
        payload.put("health", health);
        payload.put("heartbeat_interval_s", heartbeatIntervalSeconds);
        payload.put("tools", tools);
        payload.put("registrations", registrations);
        return payload;
    }

    private void send(String path, Map<String, Object> payload) throws IOException, InterruptedException {
        URI uri = URI.create(registryUrl.endsWith("/") ? registryUrl.substring(0, registryUrl.length() - 1) + path : registryUrl + path);
        HttpRequest request = HttpRequest.newBuilder(uri)
            .timeout(Duration.ofSeconds(10))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(toJson(payload)))
            .build();
        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new IOException("Registry responded with status " + response.statusCode() + ": " + response.body());
        }
    }

    private String toJson(Map<String, Object> payload) throws JsonProcessingException {
        return MAPPER.writeValueAsString(payload);
    }
}
